### PR TITLE
docs: update CICD guidelines

### PR DESCRIPTION
**Pull request by hunter**
A few tweaks to the CICD section to improve basic supply chain security and some minor cleanup. Changes include:

* Shifting code review to earlier in the process when code is committed to avoid malicious code in trunk (and stronger audit). Also avoids confusion with deploy gating which breaks Continuous Deployment.
* Recommending run CI as part of in the code review process for better coverage during approval decisions.
* Referenced Continuous Delivery and Deployment to avoid confusion that are requiring Deployment only.
* Added a reference to SHIP since it is the primary tool for Govtech and agencies.
* Capitalised "Integration" and "Deployment" for better reference to the name/abbrev

### Questions

* Is adding Delivery/Deployment too pedantic? Perhaps overkill for a guidelines but will be a common question in policy. Easier to keep to avoid future questions?
* Similar question for capitalisation for "Continuous Integration" and "Continuous Delivery". Overkill?
* These guidelines are basic but useful. Do we want to further add other guidelines to improve supply chain security (protected branches, artefact signing, etc) or will that go too far to being prescriptive?

### CODE REVIEW

File: continuous-integration-and-deployment.md [hunk 1]
```diff
@@ -1,12 +1,12 @@
-# Continuous integration and deployment
+# Continuous Integration, Delivery and Deployment
 
 ### Code reviews
 
-Code changes to production must be reviewed and approved by at least one person who is not the author of the change.
+Code changes to mainline or trunk branches must be reviewed and approved by at least one person who is not the author of the change (e.g. via a Pull or Merge Request branching workflow)
```
Comment (liangyuanruo) [inline continuous-integration-and-deployment.md:5]: Good catch! Fully agree with catching issues at the point of mainline merge.

Suggest also bolding the word "must" as per convention in the rest of the doc.

File: continuous-integration-and-deployment.md [hunk 2]
```diff
@@ -15,8 +15,10 @@ These include (but are not limited to):
 * End-to-end testing
 * Static code analysis
 
-### Continuous deployment
+It is recommended that output from Continuous Integration tests are utilised to enhance and gate the Code review process.
```
Comment (liangyuanruo) [inline continuous-integration-and-deployment.md:18]: Fully agree here; I'm however also curious for your motivation to include this. Do you think that developers are inclined to bypass their own CI processes?

**Review by liangyuanruo — APPROVED** [review]

**Comment by liangyuanruo** [comment]
On being overly pedantic, I think it makes sense to stick with "Continuous Integration" and "Continuous Delivery" as widely-understood industry terms; I think a key consideration is how the terms we choose are going to be interpreted outside the technical committee by other parties such as the project team or auditors.

On the point on being overly prescriptive for other security policies (protected branches, artefact signing, etc) I think it depends heavily on the _universality_ of the practice as well as the _costs imposed_ on the project team.

Using your examples, I think branch protection is very reasonable to introduce as a universal best practice regardless of the project, and is already implied by the current document when requiring a second approved before merging code.

For artifact signing, it's mostly applicable to external package distribution, so perhaps that can be scoped as such? It may also not be as pressing unless our org is doing this widely?

File: continuous-integration-and-deployment.md (hunk 1 above)
Comment (hunter) [reply continuous-integration-and-deployment.md:5]: Good spot. Suggestion added (if you're not aware, GH can do an inline patch as part of a comment)

```suggestion
Code changes to mainline or trunk branches **must** be reviewed and approved by at least one person who is not the author of the change (e.g. via a Pull or Merge Request branching workflow)
```

File: continuous-integration-and-deployment.md (hunk 2 above)
Comment (hunter) [reply continuous-integration-and-deployment.md:18]: Not so much that developers will bypass (although there can be a tendency to drop if too slow) but more for helping make better informed decisions when approving a PR.

In Govtech, we've been pushing for agencies to move away from Change Advisory Boards (CABs) to using PRs as a means of to accelerate change management. The PR can become audit evidence for proof that automated tests have run and manual review/approval by a quorum (not just for code but docs/arch... and of course policy writing ;)

Comment (liangyuanruo) [reply continuous-integration-and-deployment.md:18]: I see! Almost forgot about that - change management via CABs is probably far too slow. I'm wondering if we should add another section to set expectations around team organisation. I'll file an issue.

**Pull request merged**